### PR TITLE
Added `take_last` to EVC

### LINKED ISSUE #543: Redeploy of an EVC no longer fixes inconsistencies
Hi,

In the past we use redeploy feature to run a fresh setup of the EVC paths, and use it a lot in situations of problems like the EVC is having inconsistencies, the user is complaining about not forwarding, etc.

It happens that on 2023.2 we no longer can count on the redeploy feature to restart the deployment of the EVC:

```
> curl -X GET -u fulano -H 'Content-type: application/json' "https://xxxx/api/kytos/mef_eline/v2/evc/" > /tmp/evcs
> cat /tmp/evcs | jq -r '.["5679bb42a3a146"].current_path[]|.endpoint_a.id + " " + .endpoint_b.id + " " + (.metadata|tostring)'
00:00:00:00:00:xx:xx:03:1 00:00:00:00:00:xx:xx:04:1 {"s_vlan":{"tag_type":"vlan","value":15}}
00:00:00:00:00:xx:xx:10:25 00:00:00:00:00:xx:xx:04:32 {"s_vlan":{"tag_type":"vlan","value":1120}}
00:00:00:00:00:xx:xx:10:31 00:00:00:00:00:xx:xx:16:31 {"s_vlan":{"tag_type":"vlan","value":104}}
00:00:00:00:00:xx:xx:15:3 00:00:00:00:00:xx:xx:16:3 {"s_vlan":{"tag_type":"vlan","value":5}}
> curl -X PATCH -u fulano -H 'Content-type: application/json' "https://xxxx/api/kytos/mef_eline/v2/evc/{5679bb42a3a146}/redeploy"
Enter host password for user 'fulano':
{"response":"Circuit 5679bb42a3a146 redeploy received."}

> curl -X GET -u fulano -H 'Content-type: application/json' "https://xxxx/api/kytos/mef_eline/v2/evc/" > /tmp/evcs
> cat /tmp/evcs | jq -r '.["5679bb42a3a146"].current_path[]|.endpoint_a.id + " " + .endpoint_b.id + " " + (.metadata|tostring)'
00:00:00:00:00:xx:xx:03:1 00:00:00:00:00:xx:xx:04:1 {"s_vlan":{"tag_type":"vlan","value":15}}
00:00:00:00:00:xx:xx:10:25 00:00:00:00:00:xx:xx:04:32 {"s_vlan":{"tag_type":"vlan","value":1120}}
00:00:00:00:00:xx:xx:10:31 00:00:00:00:00:xx:xx:16:31 {"s_vlan":{"tag_type":"vlan","value":104}}
00:00:00:00:00:xx:xx:15:3 00:00:00:00:00:xx:xx:16:3 {"s_vlan":{"tag_type":"vlan","value":5}}

#
# As you can see, the service VLAN were still the same.
# 

# Then we decided to redeploy the other EVC (11a6a9a8778945) which was conflicting with 5679bb42a3a146

> curl -X PATCH -u fulano -H 'Content-type: application/json' "https://xxxx/api/kytos/mef_eline/v2/evc/{11a6a9a8778945}/redeploy"
Enter host password for user 'fulano':
{"response":"Circuit 11a6a9a8778945 redeploy received."}

> curl -X GET -u fulano -H 'Content-type: application/json' "https://xxxx/api/kytos/mef_eline/v2/evc/" > /tmp/evcs
Enter host password for user 'fulano':
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1892k  100 1892k    0     0   463k      0  0:00:04  0:00:04 --:--:--  463k

> cat /tmp/evcs | jq -r '.["5679bb42a3a146"].current_path[]|.endpoint_a.id + " " + .endpoint_b.id + " " + (.metadata|tostring)'
00:00:00:00:00:xx:xx:03:1 00:00:00:00:00:xx:xx:04:1 {"s_vlan":{"tag_type":"vlan","value":15}}
00:00:00:00:00:xx:xx:10:25 00:00:00:00:00:xx:xx:04:32 {"s_vlan":{"tag_type":"vlan","value":1120}}
00:00:00:00:00:xx:xx:10:31 00:00:00:00:00:xx:xx:16:31 {"s_vlan":{"tag_type":"vlan","value":104}}
00:00:00:00:00:xx:xx:15:3 00:00:00:00:00:xx:xx:16:3 {"s_vlan":{"tag_type":"vlan","value":5}}

> cat /tmp/evcs | jq -r '.["11a6a9a8778945"].current_path[]|.endpoint_a.id + " " + .endpoint_b.id + " " + (.metadata|tostring)'
00:00:00:00:00:xx:xx:10:25 00:00:00:00:00:xx:xx:04:32 {"s_vlan":{"tag_type":"vlan","value":1120}}
00:00:00:00:00:xx:xx:10:31 00:00:00:00:00:xx:xx:16:31 {"s_vlan":{"tag_type":"vlan","value":157}}
00:00:00:00:00:xx:xx:16:29 00:00:00:00:00:xx:xx:03:29 {"s_vlan":{"tag_type":"vlan","value":3}}
00:00:00:00:00:xx:xx:03:30 00:00:00:00:00:xx:xx:03:32 {"s_vlan":{"tag_type":"vlan","value":8}}
00:00:00:00:00:xx:xx:03:32 00:00:00:00:00:xx:xx:03:30 {"s_vlan":{"tag_type":"vlan","value":6}}
```

As can be seen above, even after some redeploys on two different EVCs, the service VLAN remained the same and as so the inconsistencies on the log, impacting the user. What triggered those inconsistencies? Just redeploying the EVCs in sequencial order (spaced with 5s). I had to disable both EVCs and then enable them again.

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #543

### Summary

First iteration of `take_last`, this [PR](https://github.com/kytos-ng/kytos/pull/515) needs kytos PR to work.
Added option to EVC, `take_last` which indicates which VLAN will be taken next for `current_path` and `failover_path`

### Local Tests
Create an EVC and redeploy

### End-to-End Tests
To be added.

### CODE REVIEW

**Comment by Alopalao** [comment]
Is this feature going to be backported?

File: requirements/dev.in [hunk 1]
```diff
@@ -5,5 +5,5 @@
 #    pip-compile --output-file requirements/dev.txt requirements/dev.in
 #
 -e git+https://github.com/kytos-ng/python-openflow.git#egg=python-openflow
--e git+https://github.com/kytos-ng/kytos.git#egg=kytos[dev]
+-e git+https://github.com/kytos-ng/kytos.git@feature/last_tag#egg=kytos[dev]
```
Comment (Alopalao) [inline requirements/dev.in:8]: Reminder to remove this.
Also, Is this feature going to be backported?

**Review by viniarck — CHANGES_REQUESTED** [review]
@Alopalao, that's a good start. But notice that this doesn't fully meet some of the requirements:

- If you were to have a static EVC with `primary_path`, path X, and `backup_path`, path Y, since `take_last` is a global attr of the EVC and not per link, after each redeploy for example, notice that it would keep switching and going back to use the same s_vlan for each path, path X would use 1, and paty Y would use 4095. So, in this case, if a network operator wanted to force redeploy to change `primary_path` path to use a different s_vlan it would be able to accomplish that. See what I mean? The old implementation that wouldn't be a problem since it was fully rotating the allocated s_vlan per link, so it accomplished that indirectly.

You'll need to rethink and refine the proposed solution. Here a few pointers for you to also think about:

1) I believe we'll need an EVC attr to set the links s_vlan allocation strategy. 
2) I believe we'll also need to store a state on each link metadata regarding the direction (or side) the s_vlan was taken. Since `s_vlan` is already stored as a metadata, maybe we can also reserve a new metadata for this, since no other NApp will change this since `mef_eline` is also using unique link refs for now. 

#### Quick brainstorm for you to think about

See you can come up with something better:

- Maybe we can have something like `links_svlan_alloc_strategy: str` or `links_svlan_use_strategy: str` (or something along these lines) on EVC, and then this could be a enum values of supported strategies, such as `"first|last|alternating"`, where `"alternating"` would be the default on. And, then `link.choose_vlans` would take this parameter too, to take into consideration how to allocate the svlan. If `"first"` then always try allocate the next, if `"last"` then always try allocate the last, if `"alternating"` then you either pick first or last depending on the prior state that link had on its metadata. 

Another benefit of structuring this way, in addition to meeting the requirement, is that if network operators want same allocation "direction" again then they can also stick with `"first"` and `"last"` strategy, which at some point might be very value. The main reason why they want alternating for now as you remember from recent discussion is due to potentially being able to fix s_vlans conflicts, which in normal circumstances we don't expect anymore. Also, in the very far future, this could allow us to set allocate explicit s_vlans on link (I doubt we'll need to do this in the next years, but at least it'd wouldn't be too far away, and then the strategy in this case could be `"explicit|null|None"`)

Finally, once you hash out the final solution, I'll recommend that you summarize it in the issue, and then let's see if it meets the requirement and then go for it. OK?

File: requirements/dev.in (hunk 1 above)
Comment (viniarck) [reply requirements/dev.in:8]: Yes, the underlying issue has the `backport_20232` and `backport_20241` labels.

Comment (viniarck) [reply requirements/dev.in:8]: btw, I've just merged the related PR https://github.com/kytos-ng/kytos/pull/515. Nicely done.

**Comment by viniarck** [comment]
> @Alopalao, that's a good start. But notice that this doesn't fully meet some of the requirements:

...

#### Quick brainstorm for you to think about

See you can come up with something better:

- Maybe we can have something like `links_svlan_alloc_strategy: str` or `links_svlan_use_strategy: str` (or something along these lines) on EVC, and then this could be a enum values of supported strategies, such as `"first|last|alternating"`, where `"alternating"` would be the default on. And, then `link.choose_vlans` would take this parameter too, to take into consideration how to allocate the svlan. If `"first"` then always try allocate the next, if `"last"` then always try allocate the last, if `"alternating"` then you either pick first or last depending on the prior state that link had on its metadata. 

Another benefit of structuring this way, in addition to meeting the requirement, is that if network operators want same allocation "direction" again then they can also stick with `"first"` and `"last"` strategy, which at some point might be valuable to then, and code maintenance-wise that'd be reasonable to maintain. The main reason why they want alternating for now as you remember from recent discussion is due to potentially being able to fix s_vlans conflicts, which in normal circumstances we don't expect anymore. Also, in the very far future, this could allow us to set allocate explicit s_vlans on link (I doubt we'll need to do this in the next years, but at least it'd wouldn't be too far away, and then the global link strategy in this case could be `"explicit|null|None"`, and then we could expose another way of allowing it to be set)

Finally, once you hash out the final solution, I'll recommend that you summarize it in the issue, and then let's see if it meets the requirement and then go for it. OK?

**Comment by Alopalao** [comment]
Closing this PR in favor of [avoid VLAN](https://github.com/kytos-ng/mef_eline/pull/571)

**Pull request closed without merging**